python_modules/dagster/dagster/_core/definitions/base_asset_graph.py

```diff
@@ -67,45 +70,195 @@ class ParentsPartitionsResult(NamedTuple):
     required_but_nonexistent_parents_partitions: AbstractSet[AssetKeyPartitionKey]
 
 
-class BaseAssetGraph(ABC):
+class BaseAssetNode(ABC):
+    key: AssetKey
+    _children: Optional[AbstractSet[Self]]
+    _parents: Optional[AbstractSet[Self]]
+
+    # Since both parent and child asset nodes contain refereneces to each other, it is impossible to
+    # construct a graph of all asset nodes with single-step construction. The nodes must first be
+    # constructed and then `set_neighbors` must be called to bind the references.
+
+    @property
+    def children(self) -> AbstractSet[Self]:
+        if self._children is None:
+            self._neighbors_unbound_error("child", "children")
+        return self._children
+
+    @property
+    def child_keys(self) -> AbstractSet[AssetKey]:
+        if self._children is None:
+            self._neighbors_unbound_error("child", "children")
+        return {child.key for child in self._children}
+
+    def set_children(self, children: AbstractSet[Self]) -> None:
+        self._children = children
```

Comment:
Hard no on this one.

You have two options here:

1) Change BaseAssetNode to have a reference to the AssetGraph from whence it came. This would allow you to navigate up and down the tree
2) Have BaseAssetNode know only about its upstream deps (like AssetsDefinition does). Downstream deps have to come through other abstractions, like the AssetGraph or AssetGraphView.

Reply:
I recommend option 2.